File: plugins/research-suite/skills/numerical-prototype/templates/prototype_skeleton.py
"""
prototype_skeleton.py

Starting scaffold for Stage 6 numerical prototypes. Copy into src/<package>/core.py
and fill in the physics for your specific formalism.

Conventions (see _research-commons/code_architecture/jax_first_rules.md):
- State and Params are pytree-registered dataclasses
- `step` is pure, jit-able, takes (state, key, params) -> new_state
- No Python loops in the physics core; vmap for particle batching, scan for time
- PRNGKey discipline: every stochastic function takes `key` explicitly
"""

from __future__ import annotations

from dataclasses import dataclass

import jax
import jax.numpy as jnp
from jax import Array


# --- Dtype policy ------------------------------------------------------------
# Pick one and document the reason. Mixed-precision requires an explicit policy
# in a separate module-level comment.
DTYPE = jnp.float64
# Reason: the observable is a small number (O(1e-3)) accumulated over many steps,
# so float32 accumulated error exceeds the claimed precision of the prediction.


# --- Params -----------------------------------------------------------------
@jax.tree_util.register_dataclass
@dataclass(frozen=True)
class Params:
    """Physical parameters. Do not vary during integration."""
    diffusion: float = 1.0
    interaction_strength: float = 0.0  # the "new physics" term; zero recovers known limit
    dt: float = 1e-3

    @classmethod
    def default(cls) -> "Params":
        return cls()


# --- State ------------------------------------------------------------------
@jax.tree_util.register_dataclass
@dataclass
class State:
    """Simulation state. Dtype-consistent per module policy."""
    positions: Array        # shape (N, D)
    velocities: Array       # shape (N, D)
    t: Array                # scalar

    @classmethod
    def initial(cls, n_particles: int, dim: int, key: Array) -> "State":
        k1, k2 = jax.random.split(key)
        positions = jax.random.normal(k1, (n_particles, dim), dtype=DTYPE)
        velocities = jax.random.normal(k2, (n_particles, dim), dtype=DTYPE)
        return cls(positions=positions, velocities=velocities, t=jnp.asarray(0.0, dtype=DTYPE))


# --- Physics core -----------------------------------------------------------
def _single_particle_update(pos: Array, vel: Array, key: Array, params: Params) -> tuple[Array, Array]:
    """
    Per-particle update. Called via vmap from step().
    Fill in the physics from Eq. (X) of the formalism here.
    """
    # Placeholder: overdamped Brownian dynamics with an interaction stub.
    # Replace with the physics from 05_formalism.tex.
    noise = jax.random.normal(key, pos.shape, dtype=DTYPE)
    drift = -params.interaction_strength * pos  # stub interaction
    new_pos = pos + drift * params.dt + jnp.sqrt(2 * params.diffusion * params.dt) * noise
    new_vel = vel  # unused in overdamped limit; keep for state consistency
    return new_pos, new_vel


@jax.jit
def step(state: State, key: Array, params: Params) -> State:
    """
    One integration step. Pure, jit-compiled. No Python loops.
    """
    n_particles = state.positions.shape[0]
    keys = jax.random.split(key, n_particles)
    new_pos, new_vel = jax.vmap(_single_particle_update, in_axes=(0, 0, 0, None))(
        state.positions, state.velocities, keys, params
    )
    return State(
        positions=new_pos,
        velocities=new_vel,
        t=state.t + params.dt,
    )


# --- Time integration -------------------------------------------------------
def integrate(
    initial_state: State,
    key: Array,
    params: Params,
    n_steps: int,
) -> State:
    """
    Roll out the integrator for n_steps. Returns the full trajectory stacked on
    the first axis.
    """
    keys = jax.random.split(key, n_steps)

    def body(state, step_key):
        new_state = step(state, step_key, params)
        return new_state, new_state

    _, trajectory = jax.lax.scan(body, initial_state, keys)
    return trajectory


# --- Observable extractor ---------------------------------------------------
def extract_observable(trajectory: State, params: Params) -> dict:
    """
    Convert the trajectory into the predicted observable that Stage 7 will
    design a measurement for. Replace this with the problem-specific observable
    from 04_theory.md.

    Returned dict follows templates/predicted_observable.md.
    """
    # Placeholder: mean-square displacement vs time.
    msd = jnp.mean(jnp.sum(trajectory.positions ** 2, axis=-1), axis=-1)
    t = trajectory.t
    return {
        "name": "mean_square_displacement",
        "t": t,
        "values": msd,
        "units": {"t": "time", "values": "length^2"},
    }


# --- Minimal self-test ------------------------------------------------------
def _self_test() -> None:
    """Quick sanity check that the skeleton runs."""
    key = jax.random.PRNGKey(0)
    params = Params.default()
    initial = State.initial(n_particles=100, dim=3, key=key)
    trajectory = integrate(initial, jax.random.PRNGKey(1), params, n_steps=1000)
    obs = extract_observable(trajectory, params)
    print(f"trajectory shape: {trajectory.positions.shape}")
    print(f"observable '{obs['name']}' final value: {obs['values'][-1]:.4f}")


if __name__ == "__main__":
    _self_test()

File: plugins/research-suite/skills/numerical-prototype/templates/prototype_skeleton.jl
"""
prototype_skeleton.jl

Starting scaffold for Stage 6 numerical prototypes implemented in Julia. Copy
into src/<PackageName>.jl, rename `module PrototypeSkeleton` to match the
package name, and fill in the physics for your specific formalism. The
`_self_test` block below the module is for running this file standalone
(`julia prototype_skeleton.jl`) during development; once copied into a real
package, move it into test/runtests.jl per julia_first_rules.md's layout.

Conventions (see _research-commons/code_architecture/julia_first_rules.md):
- Params and State are concrete, type-stable structs
- `step!` mutates state in place; no allocations in the hot loop
- Every stochastic function takes an `AbstractRNG` explicitly
- For stiff ODEs/DAEs/PDEs, replace the manual loop with OrdinaryDiffEq.jl
"""
module PrototypeSkeleton

using Random

export Params, State, step!, integrate!, extract_observable

# --- Dtype policy -------------------------------------------------------
# Pick one and document the reason. Mixed precision requires an explicit
# policy note, same discipline as the JAX skeleton's DTYPE constant.
const T = Float64
# Reason: the observable is a small number (O(1e-3)) accumulated over many
# steps, so Float32 accumulated error exceeds the claimed precision.

# --- Params ---------------------------------------------------------------
"""Physical parameters. Do not vary during integration."""
struct Params
    diffusion::T
    interaction_strength::T   # the "new physics" term; zero recovers the known limit
    dt::T
end

Params(; diffusion=1.0, interaction_strength=0.0, dt=1e-3) =
    Params(diffusion, interaction_strength, dt)

# --- State ------------------------------------------------------------------
"""Simulation state. Type-stable, dtype-consistent per module policy."""
mutable struct State
    positions::Matrix{T}   # (D, N): one particle per column
    velocities::Matrix{T}
    t::T

    function State(positions::Matrix{T}, velocities::Matrix{T}, t::T)
        size(positions) == size(velocities) || throw(DimensionMismatch(
            "positions $(size(positions)) and velocities $(size(velocities)) must match"))
        new(positions, velocities, t)
    end
end

function State(n_particles::Int, dim::Int, rng::AbstractRNG)
    positions = randn(rng, T, dim, n_particles)
    velocities = randn(rng, T, dim, n_particles)
    State(positions, velocities, zero(T))
end

# --- Physics core -------------------------------------------------------
"""
    step!(state, params, rng)

One integration step, mutating `state` in place. Type-stable, allocation-free
in the hot loop. Fill in the physics from Eq. (X) of the formalism here.
"""
function step!(state::State, params::Params, rng::AbstractRNG)
    # Placeholder: overdamped Brownian dynamics with an interaction stub.
    # Replace with the physics from 05_formalism.tex.
    noise_scale = sqrt(2 * params.diffusion * params.dt)
    @inbounds for i in axes(state.positions, 2)
        for d in axes(state.positions, 1)
            drift = -params.interaction_strength * state.positions[d, i]
            state.positions[d, i] += drift * params.dt + noise_scale * randn(rng, T)
        end
    end
    state.t += params.dt
    return state
end

# --- Time integration -----------------------------------------------------
"""
    integrate!(state, params, n_steps, rng) -> trajectory

Roll out `n_steps`, mutating `state` to its final value (the `!` suffix marks
this) and returning the stacked trajectory (a State per step). For stiff
governing equations, replace this loop with `OrdinaryDiffEq.jl`'s
`solve(ODEProblem(...))` instead.
"""
function integrate!(state::State, params::Params, n_steps::Int, rng::AbstractRNG)
    trajectory = Vector{State}(undef, n_steps)
    for k in 1:n_steps
        step!(state, params, rng)
        trajectory[k] = State(copy(state.positions), copy(state.velocities), state.t)
    end
    return trajectory
end

# --- Observable extractor --------------------------------------------------
"""
    extract_observable(trajectory, params) -> NamedTuple

Convert the trajectory into the predicted observable that Stage 7 will design
a measurement for. Replace with the problem-specific observable from
04_theory.md. Returns the identity/values subset of the schema in
templates/predicted_observable.md; combine with uncertainty and units via
scripts/observable_extractor.py's ObservableBuilder before the Stage 6 write-up.
"""
function extract_observable(trajectory::Vector{State}, params::Params)
    t = [s.t for s in trajectory]
    msd = [sum(abs2, s.positions) / size(s.positions, 2) for s in trajectory]
    return (name="mean_square_displacement", t=t, values=msd,
            units=(t="time", values="length^2"))
end

end # module

# --- Minimal self-test -------------------------------------------------
using .PrototypeSkeleton
using Random

"""Runnable self-check: `julia prototype_skeleton.jl`."""
function _self_test()
    rng = Random.MersenneTwister(0)
    params = Params()
    state = State(100, 3, rng)
    trajectory = integrate!(state, params, 1000, rng)
    obs = extract_observable(trajectory, params)
    println("trajectory length: ", length(trajectory))
    println("observable '", obs.name, "' final value: ", round(obs.values[end], digits=4))
    @assert length(obs.t) == length(trajectory)
    @assert all(isfinite, obs.values)
end

if abspath(PROGRAM_FILE) == @__FILE__
    _self_test()
end
